### PR TITLE
[GEOS-11330] OAuth2 kid verification should be optional

### PR DESCRIPTION
**Pull request by afabiani**
[GEOS-11330](https://osgeo-org.atlassian.net/browse/GEOS-11330) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

References: https://osgeo-org.atlassian.net/browse/GEOS-11330

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java [hunk 1]
```diff
@@ -35,7 +36,9 @@ public ValidatingOAuth2RestTemplate(
             String jwkUri,
             OpenIdConnectFilterConfig config) {
         super(resource, context);
-        if (jwkUri != null) this.store = new JwkTokenStore(jwkUri);
+        if (jwkUri != null) {
+            this.store = new JwkTokenStore(jwkUri);
+        }
```
Comment (aaime) [inline src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java:41]: Unnecessary change?

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/test/java/org/geoserver/security/oauth2/OpenIdConnectIntegrationTest.java [hunk 2]
```diff
@@ -78,7 +78,7 @@ public static void beforeClass() throws Exception {
                                         .withStatus(200)
                                         .withHeader(
                                                 "Content-Type", MediaType.APPLICATION_JSON_VALUE)
-                                        .withBodyFile("jkws.json")));
+                                        .withBodyFile("jwks.json")));
```
Comment (aaime) [inline src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/test/java/org/geoserver/security/oauth2/OpenIdConnectIntegrationTest.java:81]: Corrects the name here but the test resource modified is still named jkws.json? Is it used?

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java [hunk 3]
```diff
@@ -109,7 +112,13 @@ private void validate(OAuth2AccessToken token) {
             String idToken = (String) maybeIdToken;
             setAsRequestAttribute(OpenIdConnectAuthenticationFilter.ID_TOKEN_VALUE, idToken);
             // among other things, this verifies the token
-            if (store != null) store.readAuthentication(idToken);
+            if (store != null) {
+                try {
+                    store.readAuthentication(idToken);
+                } catch (InvalidTokenException e) {
+                    LOGGER.warning("Failed to validate ID token: " + e.getMessage());
+                }
```
Comment (aaime) [inline src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java:120]: What if the validation failed for something other than the kid verification? This seems to be ignoring all other sorts of validity issues?

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/test/java/org/geoserver/security/oauth2/OpenIdConnectIntegrationTest.java (hunk 2 above)
Comment (afabiani) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/test/java/org/geoserver/security/oauth2/OpenIdConnectIntegrationTest.java:81]: The resource has been renamed as well and added a test case.

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java (hunk 3 above)
Comment (afabiani) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java:120]: This is actually the fix, the KID verification is optional

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java (hunk 1 above)
Comment (afabiani) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java:41]: This is actually the fix, the KID verification is optional

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java (hunk 3 above)
Comment (aaime) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java:120]: Is KID verification the only possible reason to get an InvalidTokenException? Or will we ignore also verification failures that should be considered?

Comment (afabiani) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java:120]: Most likely this is the reason, other reasons might be related to the impossibility to decrypt due to cypher algorythm. In any case the verification is optional for the protocol.
In order to intercept the specific one what we can do is to inspect the exception message probably even if, generally speaking, that shouldn't be a blocking error at all.

What about having an option on the GUI in order to enforce the velidation or not instead?

Comment (aaime) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java:120]: I agree that sounds like a safer approach (let the admin decide whether to verify or not)

Comment (afabiani) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/ValidatingOAuth2RestTemplate.java]: Done

File: src/community/security/oauth2-openid-connect/oauth2-openid-connect-web/src/test/java/org/geoserver/web/security/oauth2/OpenIdConnectAuthProviderPanelTest.java [hunk 4]
```diff
@@ -95,6 +95,28 @@ public void testPostRedirectUri() throws Exception {
                 "http://localhost:8080/post/redirect");
     }
 
+    @Test
+    public void testenforceTokenValidation() throws Exception {
```
[the remark this answers is not in the document]
Comment (afabiani) [reply src/community/security/oauth2-openid-connect/oauth2-openid-connect-web/src/test/java/org/geoserver/web/security/oauth2/OpenIdConnectAuthProviderPanelTest.java:99]: Fixed thanks

**Comment by aaime** [comment]
Thanks!

**Comment by aaime** [comment]
The backport to `2.24.x` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit 4d89ccee81959cbc375b5578d3758de2ab13e28a is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
[backport-7470-to-2.24.x 6acd9dcdc1] [GEOS-11330] kid verification should be optional
 Author: afabiani <alessio.fabiani@geosolutionsgroup.com>
 Date: Mon Mar 11 13:19:27 2024 +0100
 3 files changed, 25 insertions(+), 3 deletions(-)
[backport-7470-to-2.24.x 878b7baa4a]  - Introducing the Enforce Token Validation panel option in order to make this available by the Admin
 Author: afabiani <alessio.fabiani@geosolutionsgroup.com>
 Date: Tue Mar 19 19:03:32 2024 +0100
 9 files changed, 70 insertions(+), 4 deletions(-)

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.24.x 2.24.x
# Navigate to the new working tree
cd .worktrees/backport-2.24.x
# Create a new branch
git switch --create backport-7470-to-2.24.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick ecb4184a5065be3bec7be8cbc043f3f5f37b7b4f,d985ca9ba08c43780ad270faee4584b42085d279,4d89ccee81959cbc375b5578d3758de2ab13e28a,91ded7c5685e2154a53b25e5b7bdc1c0871107d4
# Push it to GitHub
git push --set-upstream origin backport-7470-to-2.24.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.24.x
```
Then, create a pull request where the `base` branch is `2.24.x` and the `compare`/`head` branch is `backport-7470-to-2.24.x`.
